File: apps/windows/TeslaSync.App/shared-surfaces/Range.Model.cs
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// Which of the two Tesla range estimates the surface treats as "the" range — the native analogue of the web
/// <c>RangeType = 'rated' | 'ideal'</c> union (web/src/lib/preferredRange.ts) backing the user's
/// <c>preferred_range</c> General-Settings preference.
/// </summary>
public enum RangeType
{
    /// <summary>EPA-style rated range (web <c>rated</c>) — the fallback when the preference is missing.</summary>
    Rated,

    /// <summary>Ideal range (web <c>ideal</c>).</summary>
    Ideal,
}

/// <summary>
/// The minimal vehicle/charge state snapshot the <see cref="Range"/> surface reads — the native analogue of the
/// web <c>PreferredRangeFields</c> prop (<c>{ rated_range?, ideal_range? }</c>, both SI metres) that
/// <c>&lt;Range /&gt;</c> consumes (web/src/components/data-display/format/Range.tsx). Both fields are nullable and
/// stay null when the source did not report them, so the projection renders the em-dash empty display rather than
/// a fabricated value — exactly as the web component renders <c>—</c> when <c>usePreferredRange(state).meters</c>
/// is null.
/// </summary>
/// <param name="RatedRangeMeters">Rated range in SI metres (web <c>rated_range</c>), or null when absent.</param>
/// <param name="IdealRangeMeters">Ideal range in SI metres (web <c>ideal_range</c>), or null when absent.</param>
public readonly record struct RangeState(double? RatedRangeMeters, double? IdealRangeMeters);

/// <summary>
/// Canonical metadata for the Range surface — the native analogue of the module-level contract in
/// <c>web/src/components/data-display/format/Range.tsx</c>. The web component is a pure presentational leaf (a
/// <c>&lt;span&gt;</c> that renders <c>formatDistance(meters, { precision })</c>, or <c>—</c> when the preferred
/// range is null), so this carries only the diagnostics slug, the automation id, the web <c>precision = 0</c>
/// default, and the i18n keys the companion <c>useRangeLabel</c> resolves through.
/// </summary>
public static class RangeRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "Range";

    /// <summary>
    /// The root automation id the view stamps on itself. The web component declares no <c>data-testid</c> (it is
    /// an anonymous inline span), so this is the native-only stable hook UI-automation tests target.
    /// </summary>
    public const string RootAutomationId = "range";

    /// <summary>Default decimal precision for the rendered distance (web <c>precision = 0</c>).</summary>
    public const int DefaultPrecision = 0;

    /// <summary>The i18n key for the rated-range label (web <c>t('common.ratedRange')</c>).</summary>
    public const string RatedLabelKey = "common.ratedRange";

    /// <summary>The English fallback for the rated-range label (web default).</summary>
    public const string RatedLabelDefault = "Rated Range";

    /// <summary>The i18n key for the ideal-range label (web <c>t('common.idealRange')</c>).</summary>
    public const string IdealLabelKey = "common.idealRange";

    /// <summary>The English fallback for the ideal-range label (web default).</summary>
    public const string IdealLabelDefault = "Ideal Range";

    /// <summary>The i18n key for the Narrator "no value" announcement of the empty display.</summary>
    public const string NoValueKey = "common.noData";

    /// <summary>The English fallback for the Narrator "no value" announcement.</summary>
    public const string NoValueDefault = "No data available";
}

/// <summary>
/// Pure projection of the Range readout's render inputs — the native port of the web component body
/// (web/src/components/data-display/format/Range.tsx) plus its <c>useRangeLabel</c> companion. It selects the
/// preferred range value (the port of <c>selectPreferredRange</c> in web/src/lib/preferredRange.ts — defaulting to
/// rated when the preference is missing), formats it through the shared, unit-tested
/// <see cref="UnitFormatters.FormatDistance(double?, UnitPref, int?)"/> (the verified 1:1 of the web
/// <c>formatDistance(meters, { precision })</c>, which itself renders the em dash for null/NaN), and resolves the
/// rated/ideal <see cref="Label"/> through the i18n facade. Kept static and side-effect-free so the adapter is
/// unit-testable without a view-model or a UI thread; the <see cref="RangeViewModel"/> and the WinUI view both
/// render from it.
/// </summary>
public readonly record struct RangeProjection
{
    private RangeProjection(double? meters, RangeType source, bool hasValue, string value, string label, string accessibleName)
    {
        Meters = meters;
        Source = source;
        HasValue = hasValue;
        Value = value;
        Label = label;
        AccessibleName = accessibleName;
    }

    /// <summary>The selected range value in SI metres (web <c>usePreferredRange(state).meters</c>), or null.</summary>
    public double? Meters { get; }

    /// <summary>Which range field was selected (web <c>source</c>), honoring the preferred-range preference.</summary>
    public RangeType Source { get; }

    /// <summary>
    /// True when a finite range value is present. Gates the friendly empty display: false renders the em dash, the
    /// native analogue of the web component returning <c>&lt;span&gt;—&lt;/span&gt;</c> for a null range.
    /// </summary>
    public bool HasValue { get; }

    /// <summary>
    /// The visible readout: the formatted distance honoring the unit + precision preference, or the em-dash empty
    /// display (<see cref="UnitFormatters.DefaultEmptyDisplay"/>) when no value is present.
    /// </summary>
    public string Value { get; }

    /// <summary>
    /// The localized rated/ideal range label (web <c>useRangeLabel</c>). Exposed so a host can render the label
    /// separately from the value and so the surface can build a meaningful Narrator name.
    /// </summary>
    public string Label { get; }

    /// <summary>
    /// The surface's accessible name (Narrator). Combines <see cref="Label"/> with the value — or with the
    /// localized "no value" announcement when empty — so the screen reader hears "Rated Range: 410 km" rather than
    /// a bare em dash.
    /// </summary>
    public string AccessibleName { get; }

    /// <summary>
    /// Project the render inputs. Selects rated vs ideal from <paramref name="preferredRange"/> (defaulting to
    /// rated, matching the web <c>selectPreferredRange</c> fallback), formats the selected metres through
    /// <see cref="UnitFormatters.FormatDistance(double?, UnitPref, int?)"/> with a non-negative precision, and
    /// resolves the rated/ideal label through <paramref name="localizer"/>.
    /// </summary>
    /// <param name="state">The vehicle/charge snapshot (web <c>state</c> prop); null while loading/absent.</param>
    /// <param name="preferredRange">The user's preferred-range preference (web <c>useSettings().rangeType</c>).</param>
    /// <param name="units">The user's display-unit preference (web <c>useUnits().unitPrefs</c>).</param>
    /// <param name="precision">The decimal precision (web <c>precision</c>); negative is treated as the default.</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    /// <returns>The render-ready projection.</returns>
    public static RangeProjection Project(
        RangeState? state,
        RangeType preferredRange,
        UnitPref units,
        int precision,
        ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(units);
        ArgumentNullException.ThrowIfNull(localizer);

        bool ideal = preferredRange == RangeType.Ideal;
        double? meters = ideal ? state?.IdealRangeMeters : state?.RatedRangeMeters;
        RangeType source = ideal ? RangeType.Ideal : RangeType.Rated;

        int safePrecision = precision < 0 ? RangeRegistration.DefaultPrecision : precision;

        // Web: meters == null ? '—' : formatDistance(meters, { precision }). FormatDistance itself returns the em
        // dash for null/NaN, so a single call reproduces both branches.
        string value = UnitFormatters.FormatDistance(meters, units, safePrecision);
        bool hasValue = meters is { } m && !double.IsNaN(m) && !double.IsInfinity(m);

        // Web useRangeLabel: t(`common.${labelKey}`, defaultLabel).
        string label = ideal
            ? localizer.GetString(RangeRegistration.IdealLabelKey, RangeRegistration.IdealLabelDefault)
            : localizer.GetString(RangeRegistration.RatedLabelKey, RangeRegistration.RatedLabelDefault);

        string announced = hasValue
            ? value
            : localizer.GetString(RangeRegistration.NoValueKey, RangeRegistration.NoValueDefault);
        string accessibleName = $"{label}: {announced}";

        return new RangeProjection(meters, source, hasValue, value, label, accessibleName);
    }
}

/// <summary>
/// PII-safe diagnostics for the Range surface (P1/S11 diagnostics contract). The readout carries only a
/// caller-supplied range value, so the collector records only the operational <c>view.opened</c> event with the
/// surface slug — never the range, VIN or location. Thread-safe.
/// </summary>
public sealed class RangeDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">The sink the <c>view.opened</c> line is written to, or null.</param>
    public RangeDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=Range</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={RangeRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/RangeModel.kt
// Pure, framework-free model + projection for the Range shared surface — the native analogue of the data
// the web component derives before returning JSX (web/src/components/data-display/format/Range.tsx: the
// `useUnits` + `usePreferredRange` + `useRangeLabel` composition). No Compose, no Android UI, no HTTP:
// every type here is exercised by the :android:testReleaseUnitTest gate so the composable stays a thin
// render layer.
//
// The web `Range` is a tiny presentational formatter: it reads the user's distance-unit preference
// (`useUnits` → `unit_of_length`) and `preferred_range` preference (`usePreferredRange` → rated vs ideal),
// selects `state.rated_range`/`state.ideal_range` (SI metres), and renders either the formatted distance
// or an em dash when the value is missing. `useRangeLabel` returns the localized "Rated Range"/"Ideal
// Range" label. This model reproduces that selection + formatting exactly (SI-floor, Phase-48), and folds
// in the cache-then-network lifecycle of the settings document (the genuine async dependency behind both
// hooks) so the surface can honestly render the prompt's loading / empty / error / stale / offline matrix
// without ever hiding a region.
//
// There is no native `selectPreferredRange` in the shared core (the web `lib/preferredRange.ts` helper has
// no KMP port), so its pure logic is reproduced here — the same approach the sibling surfaces take for
// their own per-surface projection logic. The settings document and the unit derivation come from the
// shared layers: `UnitPreferences.fromSettings` (the `useUnits` port) and the SI `formatDistance` (the
// `lib/unitConversion` port).
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/Range — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen is illegal in a package identifier), so the package intentionally diverges from the
// path. `MatchingDeclarationName` is suppressed for the co-located supporting types.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.range

import io.teslasync.android.components.feedback.QueryErrorKind
import io.teslasync.android.data.ErrorKind
import io.teslasync.android.data.UiState
import io.teslasync.android.data.UnitPreferences
import io.teslasync.shared.core.units.formatDistance
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull

/**
 * Canonical registry metadata for this surface — the native mirror of the web component's contract. The
 * diagnostics slug, the settings key the range-type preference reads, and the SI snapshot keys are pinned
 * here so the native and web surfaces stay in lockstep.
 */
object RangeRegistration {
    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "Range"

    /** Settings-document key holding the user's range-type preference (web `useSettings().rangeType`). */
    const val PREFERRED_RANGE_KEY: String = "preferred_range"

    /** The `preferred_range` value that selects the ideal estimate; anything else falls back to rated. */
    const val RANGE_TYPE_IDEAL: String = "ideal"

    /** Snapshot key for the rated range in SI metres (web `PreferredRangeFields.rated_range`). */
    const val RATED_RANGE_KEY: String = "rated_range"

    /** Snapshot key for the ideal range in SI metres (web `PreferredRangeFields.ideal_range`). */
    const val IDEAL_RANGE_KEY: String = "ideal_range"

    /** The em dash the web renders when the selected range is missing (`<span>—</span>`). */
    const val EMPTY_VALUE: String = "\u2014"
}

/**
 * Which of Tesla's two range estimates the user treats as "the" range — the native port of the web
 * `RangeType` union ('rated' | 'ideal'). Rated is the fallback when the preference is missing or mistyped,
 * matching the backend default the web `useSettings` applies.
 */
enum class PreferredRangeType { Rated, Ideal }

/**
 * A vehicle/charge state snapshot carrying the two range estimates in SI metres — the native port of the
 * web `PreferredRangeFields` (`{ rated_range?, ideal_range? }`). A `null` field means "not reported yet",
 * which the projection renders as the em-dash empty state, exactly like the web `meters == null` branch.
 */
data class RangeSnapshot(
    val ratedRangeMeters: Double? = null,
    val idealRangeMeters: Double? = null,
) {
    companion object {
        /**
         * Builds a snapshot from a raw vehicle/charge state object (the shape a host already holds),
         * reading the SI-metre `rated_range`/`ideal_range` fields. A null/non-object element, or a missing
         * field, yields a null estimate. Pure, so it is covered by the off-device projection test.
         */
        fun fromState(state: JsonElement?): RangeSnapshot {
            val obj = state as? JsonObject ?: return RangeSnapshot()
            return RangeSnapshot(
                ratedRangeMeters = obj.metres(RangeRegistration.RATED_RANGE_KEY),
                idealRangeMeters = obj.metres(RangeRegistration.IDEAL_RANGE_KEY),
            )
        }

        private fun JsonObject.metres(key: String): Double? = (this[key] as? JsonPrimitive)?.doubleOrNull
    }
}

/**
 * The mutually-exclusive render surface the Range card draws. [Content] and [Empty] reproduce the web's
 * two visible branches (the formatted value vs the em dash); [Loading] and [Error] surface the genuine
 * cold-start and hard-failure states of the settings document the unit + range-type preferences come from.
 */
enum class RangePhase {
    /** First settings load with nothing cached — render skeleton chrome (never a blank box). */
    Loading,

    /** A range value is available and formatted in the user's units — render it. */
    Content,

    /** Settings resolved but the selected range is missing (web `meters == null`) — render the em dash. */
    Empty,

    /** Settings failed with nothing cached to fall back on — render a classified error with retry. */
    Error,
}

/**
 * The immutable, render-ready projection the composable draws — everything the web `Range` + `useRangeLabel`
 * fold together: the resolved [rangeType] (for the label), the formatted [valueText] (null unless
 * [RangePhase.Content]), and the cache-then-network freshness envelope ([stale]/[offline]/[refreshing] +
 * [errorKind]) so the surface honestly flags last-known data instead of presenting it as live. Pure data
 * so [RangeProjection] is unit-tested without a UI host.
 *
 * @property stale cached prefs are past their TTL and a refresh is in flight (no failure yet).
 * @property offline cached prefs are shown because a refresh failed (network unreachable / "last known").
 * @property freshnessStamp the `fetchedAt` of the shown prefs; keys the stale auto-refresh effect.
 */
data class RangeDisplay(
    val phase: RangePhase,
    val rangeType: PreferredRangeType,
    val valueText: String? = null,
    val stale: Boolean = false,
    val offline: Boolean = false,
    val refreshing: Boolean = false,
    val errorKind: ErrorKind? = null,
    val httpStatus: Int? = null,
    val freshnessStamp: Long? = null,
) {
    /** The value to draw: the formatted distance for content, else the em dash (web `<span>—</span>`). */
    val displayValue: String get() = valueText ?: RangeRegistration.EMPTY_VALUE

    /** True when a freshness chip (stale or offline) should be shown over the cached value. */
    val showFreshnessChip: Boolean get() = stale || offline

    /** True when a retry affordance should be offered (the hard-error surface). */
    val canRetry: Boolean get() = phase == RangePhase.Error
}

/**
 * Pure projection + selection logic for the Range surface — the native port of the web
 * `selectPreferredRange` helper plus the `useUnits`/`useRangeLabel` derivation.
 */
object RangeProjection {
    private const val HTTP_UNAUTHORIZED = 401
    private const val HTTP_FORBIDDEN = 403
    private const val HTTP_NOT_FOUND = 404

    /**
     * Reads the user's range-type preference from the settings document — the native port of the web
     * `useSettings().rangeType` (`s.preferred_range`). `"ideal"` selects ideal; any other value, a missing
     * key, or an unresolved document falls back to rated (the web `rangeType === 'ideal' ? 'ideal' : 'rated'`).
     */
    fun rangeTypeOf(settings: JsonElement?): PreferredRangeType {
        val obj = settings as? JsonObject ?: return PreferredRangeType.Rated
        val raw = (obj[RangeRegistration.PREFERRED_RANGE_KEY] as? JsonPrimitive)?.contentOrNull
        return if (raw == RangeRegistration.RANGE_TYPE_IDEAL) PreferredRangeType.Ideal else PreferredRangeType.Rated
    }

    /**
     * Picks the preferred range value (SI metres) from a [snapshot] — the native port of the web
     * `selectPreferredRange`'s `type === 'ideal' ? state?.ideal_range : state?.rated_range`. A null snapshot
     * or a missing estimate yields null, which the projection renders as the empty em dash.
     */
    fun selectMetres(
        snapshot: RangeSnapshot?,
        type: PreferredRangeType,
    ): Double? =
        when (type) {
            PreferredRangeType.Ideal -> snapshot?.idealRangeMeters
            PreferredRangeType.Rated -> snapshot?.ratedRangeMeters
        }

    /**
     * Folds the settings [UiState] (the unit + range-type preference source), the provided [snapshot] (the
     * web `state` prop), and the display [precision] (web default 0) into the render-ready [RangeDisplay].
     *
     * Phase resolution honours both the web's two visible branches and the settings document's async
     * lifecycle: a hard settings failure with no cache → [RangePhase.Error]; a first load with nothing
     * cached → [RangePhase.Loading]; otherwise the user's prefs are available (fresh or cached) and the
     * selected range decides [RangePhase.Empty] (web `meters == null`) vs [RangePhase.Content]. The value is
     * formatted at the SI display boundary with the resolved [io.teslasync.shared.core.units.UnitPref], so a
     * km/mi preference change re-renders the value with no other change (the web `useUnits` contract).
     */
    fun project(
        settings: UiState<JsonElement>,
        snapshot: RangeSnapshot?,
        precision: Int,
    ): RangeDisplay {
        val type = rangeTypeOf(settings.data)
        val metres = selectMetres(snapshot, type)
        val prefs = UnitPreferences.fromSettings(settings.data)
        val phase =
            when {
                settings.isError -> RangePhase.Error
                settings.isLoading -> RangePhase.Loading
                metres == null -> RangePhase.Empty
                else -> RangePhase.Content
            }
        return RangeDisplay(
            phase = phase,
            rangeType = type,
            valueText = if (phase == RangePhase.Content) formatDistance(metres, prefs, precision) else null,
            stale = settings.stale && settings.errorKind == null,
            offline = settings.stale && settings.hasData && settings.errorKind != null,
            refreshing = settings.refreshing,
            errorKind = settings.errorKind,
            httpStatus = settings.httpStatus,
            freshnessStamp = settings.fetchedAt,
        )
    }

    /**
     * Maps the hard-error [display] onto the shared [QueryErrorKind] recovery bucket so the error surface
     * shows the right copy: an open breaker → [QueryErrorKind.Waiting]; a connectivity failure →
     * [QueryErrorKind.Network]; a 401/403 → [QueryErrorKind.Unauthorized]; a 404 → [QueryErrorKind.NotFound];
     * every other HTTP/decode/unknown failure → [QueryErrorKind.ServerError] with a retry affordance.
     */
    fun queryErrorKind(display: RangeDisplay): QueryErrorKind =
        when (display.errorKind) {
            ErrorKind.CircuitOpen -> QueryErrorKind.Waiting
            ErrorKind.Network, ErrorKind.Timeout -> QueryErrorKind.Network
            ErrorKind.Http ->
                when (display.httpStatus) {
                    HTTP_UNAUTHORIZED, HTTP_FORBIDDEN -> QueryErrorKind.Unauthorized
                    HTTP_NOT_FOUND -> QueryErrorKind.NotFound
                    else -> QueryErrorKind.ServerError
                }
            ErrorKind.Decode, ErrorKind.Unknown, null -> QueryErrorKind.ServerError
        }
}
